Create URI filter for the search API

jingtang10 opened the pull request:
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #951 

**Description**
Create the URI filters for the search API

Refactoring included

- Reorder test cases in `SearchTest.kt`
- Rename uri index table column name to be consistent with other tables

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Feature

**Screenshots (if applicable)**
NA

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md)
- [x] I have read [How to Contribute](https://github.com/google/android-fhir/blob/master/docs/contributing.md)
- [x] I have read the [Developer's guide](https://github.com/google/android-fhir/wiki/Developer's-Guide)
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate )
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally
- [x] I have built and run the reference app(s) to verify my change fixes the issue and/or does not break the reference app(s)

epicadk commented:
That's strange I also had to do this manually (https://github.com/google/android-fhir/pull/952/commits/b690b675362841283baede23fe7191535810e413)

jingtang10 commented:
> That's strange I also had to do this manually ([b690b67](https://github.com/google/android-fhir/commit/b690b675362841283baede23fe7191535810e413))

this is because of @kevinmost's PR #930 also updated the licence header. So if you had run spotless and added new files in a pending PR you may have needed to rerun spotless.

juanchazo reviewed (approved):
Just two minor comments.

aditya-07 reviewed (approved):
LGTM

jingtang10 commented on engine/src/main/java/com/google/android/fhir/search/filter/UriParamFilterCriterion.kt:
```diff
@@ -0,0 +1,35 @@
+import com.google.android.fhir.search.SearchQuery
+import org.hl7.fhir.r4.model.ResourceType
+
+class UriParamFilterCriterion(val parameter: UriClientParam, var value: String? = null) :
```
apologies this got merged after @aditya-07 approved this because I turned on auto-merge... will put this into my next pr (and tag you)

jingtang10 commented on engine/src/main/java/com/google/android/fhir/search/filter/UriParamFilterCriterion.kt:
```diff
@@ -0,0 +1,35 @@
+      SELECT resourceId FROM UriIndexEntity
+      WHERE resourceType = ? AND index_name = ? AND index_value = ? 
+      """,
+      listOf(type.name, parameter.paramName, value!!)
```
ditto

jingtang10 replied:
i think all filters actually use !! to check non-null.
